torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1092,7 +1092,14 @@ void ProcessGroupNCCL::shutdown(c10::optional<std::string> reason) {
   LOG(INFO) << logPrefix()
             << "Launching ProcessGroupNCCL abort asynchrounously.";
   std::future<bool> fut = std::async(
-      std::launch::async, [this, &reason]() { return this->abort(reason); });
+      std::launch::async, [this, &reason]() {
+#ifdef ENABLE_NCCL_ERROR_CHECKING
+    if (ncclCommWatchdogThread_.joinable()) {
+      ncclCommWatchdogThread_.join();
```

Comment:
This looks good to me but also Note we have monitor thread too, we can also wait for its termination here

Reply:
note the comment above.  I'm not seeing the full picture at the moment but is this change completely thought through?

"""
// Don't join threads here since the purpose of this method is to abort all
  // communicators and signal the threads to exit. Joining on the threads could
  // potentially block and hence avoid it in this method.
"""